### ISSUE TITLE
Due dates deleted during page reload

### ISSUE THREAD

**Issue by doitux** [nextcloud/tasks#167]
### Steps to reproduce
1. Add a task
2. Select the new task and select a due date
3. Reload the page in the browser 

### Expected behaviour
The due date should stay as before

### Actual behaviour
The recent added due date is deleted

## Server configuration detail

**Operating system:** FreeBSD 10.3-RELEASE-p17 hostBSD 10.3-RELEASE-p17 #6: Mon Mar 13 00:47:05 CET 2017     root@mywebsite.de:/usr/obj/usr/src/sys/DMRKERNEL amd64

**Webserver:** Apache/2.4.25 (FreeBSD) OpenSSL/1.0.2k mod_fcgid/2.3.9 (cgi-fcgi)

**Database:** mysql 5.6.35

**PHP version:** 7.0.16
Modules loaded: Core, date, libxml, pcre, Reflection, SPL, hash, session, cgi-fcgi, standard, apcu, calendar, ctype, curl, dom, fileinfo, filter, ftp, gd, gettext, mysqlnd, iconv, intl, imagick, imap, json, mbstring, mcrypt, mysqli, openssl, PDO, posix, SimpleXML, sqlite3, tokenizer, xml, xmlwriter, zip, zlib, pdo_mysql, pdo_sqlite, Phar, soap, xmlreader, xsl, ionCube Loader, Zend OPcache

**Nextcloud version:** 13.0.2 - 13.0.2.1

<details><summary>Signing status</summary>

Array
</details>

<details><summary>List of activated apps</summary>

```
Enabled:
 - calendar: 1.6.1
 - contacts: 2.1.3
 - dav: 1.4.6
 - federatedfilesharing: 1.3.1
 - federation: 1.3.0
 - files: 1.8.0
 - files_sharing: 1.5.0
 - files_texteditor: 2.5.1
 - files_trashbin: 1.3.0
 - files_versions: 1.6.0
 - firstrunwizard: 2.2.1
 - issuetemplate: 0.3.0
 - lookup_server_connector: 1.1.0
 - news: 12.0.1
 - nextcloud_announcements: 1.2.0
 - notifications: 2.1.2
 - oauth2: 1.1.0
 - password_policy: 1.3.0
 - provisioning_api: 1.3.0
 - serverinfo: 1.3.0
 - survey_client: 1.1.0
 - systemtags: 1.3.0
 - tasks: 0.9.6
 - twofactor_backupcodes: 1.2.3
 - updatenotification: 1.3.0
 - workflowengine: 1.3.0
Disabled:
 - activity
 - admin_audit
 - comments
 - encryption
 - files_external
 - files_pdfviewer
 - files_videoplayer
 - gallery
 - logreader
 - sharebymail
 - theming
 - user_external
 - user_ldap

```
</details>

<details><summary>Configuration (config/config.php)</summary>

```
{
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "mycloud.mywebsite.de"
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "overwrite.cli.url": "https:\/\/mycloud.mywebsite.de",
    "dbtype": "mysql",
    "version": "13.0.2.1",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "maintenance": false,
    "mysql.utf8mb4": true,
    "loglevel": 0,
    "memcache.local": "\\OC\\Memcache\\APCu",
    "updater.secret": "***REMOVED SENSITIVE VALUE***",
    "theme": ""
}
```

**Comment by raimund-schluessler** [comment]
Please report this in the https://github.com/nextcloud/server repository. This app is just a client and is not involved in the update process.

**Comment by doitux** [comment]
Please reopen this issue because i realized this also happens when just do a reload of the page in the browser directly after selecting a due date!

**Comment by raimund-schluessler** [comment]
The tasks are currently not immediately synced to the server. Does this also happen if you wait for some seconds?

**Comment by doitux** [comment]
I tested waiting 20 Minutes. But it does not help. After reloading the page the due date is deleted.

**Comment by raimund-schluessler** [comment]
You need to open the Web-Console. For Firefox the shortcut is Ctrl+Shift+K. For Chrome you have to press Ctrl+Shift+I and then click on console.

When you then set a due date, you should see a request to the server. e.g. 
<img width="907" alt="firefoxrequest" src="https://user-images.githubusercontent.com/2496460/39406889-46a57de8-4bbe-11e8-95c0-e6aa3900bb3f.png">

When you click on this request, you should see what date was send.

**Comment by doitux** [comment]
Yes i can see a PUT request directly after selecting the due date: 
https://mycloud.mywebsite.de/remote.php/dav/calendars/doitux/aufgaben/Nextcloud-kctcjpashyf1lhk0lbch1j.ics

But anyway after reloading the page the due date is deleted. Can you reproduce this issue?

**Comment by raimund-schluessler** [comment]
> Can you reproduce this issue?

No, the dates are saved correclty for me.

Could you please check the request and look if the due date is given:

```BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Nextcloud Tasks v0.9.6
BEGIN:VTODO
CREATED:20180227T103432
DTSTAMP:20180429T225107
LAST-MODIFIED:20180429T225107
UID:gc6akb6ojdd
SUMMARY:Tasktitle
PRIORITY:0
PERCENT-COMPLETE:0
X-OC-HIDESUBTASKS:0
DUE:20180429T230000
END:VTODO
END:VCALENDAR
```

There should be a line starting with `DUE` e.g. `DUE:20180429T230000` as in the above given task.
Also, are other changes to the task saved?

**Comment by doitux** [comment]
I've selected the due date of new task to 2018-05-17 and i found this data in the request: 
`BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Nextcloud Tasks v0.9.6
BEGIN:VTODO
CREATED:20180430T082557
DTSTAMP:20180430T082724
LAST-MODIFIED:20180430T082724
UID:2np3fmmksnp
SUMMARY:rssdf
PRIORITY:0
PERCENT-COMPLETE:0
X-OC-HIDESUBTASKS:0
DUE:20180517T090000
END:VTODO
END:VCALENDAR
`

**Comment by raimund-schluessler** [comment]
Yes, the server reported an error, although I don't know what could cause this. I think it is not the Tasks app causing it, but this is not sure. Maybe @georgehrke knows what could be wrong?

**Comment by doitux** [comment]
I've played around and maybe found a useful information:
When i add the task and do a page reload BEFORE i select a due date than the due date stays correctly also after another page reload.

**Comment by raimund-schluessler** [comment]
Definitely interesting, but doesn't really explain the problem. Due dates are saved correctly for me after creating a task and setting the date without any page reload.

**Comment by armaccloud** [comment]
@raimund-schluessler 
Yes, in what logs, where do I have to look?

**Comment by armaccloud** [comment]
The following errors occur when setting the due date (precondition failed)

![image](https://user-images.githubusercontent.com/9272078/39920476-7808264e-5517-11e8-950f-83d32327b422.png)

```
The following information in params
BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Nextcloud Tasks v0.9.6
BEGIN:VTODO
CREATED:20180511T123206
DTSTAMP:20180511T123220
LAST-MODIFIED:20180511T123220
UID:4fndi3vi529
SUMMARY:Test
PRIORITY:0
PERCENT-COMPLETE:0
X-OC-HIDESUBTASKS:0
DUE:20180529T130000
END:VTODO
END:VCALENDAR
```
The following response
```
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\PreconditionFailed</s:exception>
  <s:message>An If-Match header was specified, but none of the specified the ETags matched.</s:message>
  <s:header>If-Match</s:header>
</d:error>
```

**Comment by raimund-schluessler** [comment]
Yes, this is the same problem. Unfortunately I still have no idea what causes this. Could you please check if this happens also for other changes to the task, like setting the importance or the percent finished?

**Comment by armaccloud** [comment]
Ok, I have done another test
1. Created a new task, filled in all fields (start date (all day), due date (all day), priority, percentage completed, category, free field)
2. Clicked on Files App, then reopened Tasks app

**Result**
All fields are empty, except for the title

Retried this test just by refreshing the Tasks app page after creating a task, which has the same result
If I populate the same fields again on the existing task, it does save the details. So it's on the initial creation of the task that something goes wrong.

**Comment by armaccloud** [comment]
**Initial task creation (only title)**
![image](https://user-images.githubusercontent.com/9272078/39955744-b251f506-55d4-11e8-9b24-b985478d079a.png)

**Updating that task**
![image](https://user-images.githubusercontent.com/9272078/39955753-dd795580-55d4-11e8-8100-d5993bf8d3db.png)

**Initial Calendar Event creation**
![image](https://user-images.githubusercontent.com/9272078/39955773-50330b66-55d5-11e8-9a78-346a281e8874.png)

**Calendar Event Update (free text addition)**
![image](https://user-images.githubusercontent.com/9272078/39955779-6cda4f22-55d5-11e8-9663-e7450241ae2f.png)

**Comment by armaccloud** [comment]
@raimund-schluessler 
Is this all the info you need? Let me know if any more troubleshooting is necessary.

**Comment by raimund-schluessler** [comment]
Well, sadly I am out of ideas what could cause this.

**Comment by blizzz** [comment]
This is most likely a web server misconfiguration. I ran into it, too, and debugged it. I've got Apache2 and have following pieces in my configuration:

```
        …
        RequestHeader  edit "If-None-Match" "^\"(.*)-gzip\"$" "\"$1\""
        Header  edit "ETag" "^\"(.*[^g][^z][^i][^p])\"$" "\"$1-gzip\""
        …
```

Essentially, this strips off the "-gzip" suffix, which would be breaking caching. This is done when sending out the ETag header and when receiving the If-None-Match header. The tasks app is sending the "If-Match" header however which must fit to the Etag of the corresponding task. After adding it to the configuration, it works as expected:

```
        …
        RequestHeader  edit "If-None-Match" "^\"(.*)-gzip\"$" "\"$1\""
        RequestHeader  edit "If-Match" "^\"(.*)-gzip\"$" "\"$1\""
        Header  edit "ETag" "^\"(.*[^g][^z][^i][^p])\"$" "\"$1-gzip\""
        …
```

It is in Apache2 in my case, but the header modification can be done in any server.

**Comment by raimund-schluessler** [comment]
Thanks a lot @blizzz, very interesting. But I still wonder, why this is not happening for the Calendar app. The two apps seem to behave differently here. I guess this could be tackled with the move to the https://github.com/nextcloud/cdav-library in the vue branch without server reconfiguration as well.

**Comment by blizzz** [comment]
> But I still wonder, why this is not happening for the Calendar app.

Implementation details I guess. 

> without server reconfiguration as well.

If the server is configured as above, If-Match should be included to work properly. Independent from this case, there might be other apps or services that could suffer the same way. This whole thing is a hack anyhow, probably not needed with future Apache 2.6.
